### DISCUSSION TITLE
Stroke to Path?

### DISCUSSION THREAD

**Discussion by Apanatshka** [spakin/SimpInkScr#100] [Q&A] [answered]
Is there a way to call the Path -> Stroke to Path functionality?
I've found the path operations like union/difference/etc, I've found the `to_path` method which corresponds to Path -> Object to Path, but I could find stroke to path...

**Comment by spakin** [comment]
I found a partial solution for you:
```Python
for obj in selected_shapes():
    svg = svg_root.tostring()
    iobj = obj.get_inkex_object()
    new_svg = \
        inkex.command.inkscape_command(svg,
                                       select=iobj.get_id(),
                                       actions=['object-stroke-to-path'])
    loaded_svg = inkex.elements.load_svg(new_svg)
    extension.document = loaded_svg
```
The idea is to spawn a separate Inkscape (using Inkscape 1.3's `inkex.command.inkscape_command` function) that invokes the `object-stroke-to-path` action on a specific object.  (Run `inkscape --action-list` for a list of all available actions.)  The script the overwrites the existing document's SVG with the SVG returned by the child Inkscape.

The reason this is only a partial solution is that Simple Inkscape Scripting does not know that the SVG file has changed and that some new objects are available and some old objects have disappeared.  Hence, a script cannot perform further operations on an object whose stroke was converted to a path.

Perhaps a future version of Simple Inkscape Scripting will provide the internal mechanisms to launch a child Inkscape and recreate all Simple Inkscape Scripting objects upon its return, but I can't promise anything.

**Comment by Apanatshka** [comment]
Thanks, that's indeed a partial solution. In my case the full thing I wanted to do was automate the removal of the stroke from the background of many objects (since the stroke is halfway over the fill), so my manual process was to do (1) stroke-to-path, (2) ungroup, (3) difference. For now, I used an external program to automate sending the hotkeys for those three actions, and just slogged through selecting the many objects and sending those hotkeys.

It would be cool to be able to access actions from SimplInkScr directly, but if I understand you correctly, Inkscape is missing some extension API stuff to make that easily available. Thanks for your help though :)

**Reply by spakin** [reply, marked as answer]
Good news!  I believe I managed to implement a means of invoking arbitrary Inkscape actions from a Simple Inkscape Scripting script.

I introduced an `apply_action` function, which invokes `inkscape_command` as in the partial solution but then patches Simple Inkscape Scripting's state so it's aware that the document has changed from under it.

Documentation for `apply_action` is available under [Inkscape actions](https://github.com/spakin/SimpInkScr/wiki/Other-features#inkscape-actions) in the Simple Inkscape Scripting wiki.

Please download the latest version of the code from the GitHub repository and let me know if this provides the functionality you were looking for.

**Reply by Apanatshka** [reply]
Thanks Scott, it works! I managed to use the new function effectively to do my bulk operation:
```python3
for obj in selected_shapes():
	split_and_group = apply_action('object-stroke-to-path', obj)
	apply_path_operation('difference', [split_and_group[1], split_and_group[2]])
```

Side-note: the group may be broken somehow, because the following version does not work, it fails on the ungroup method...
```python3
for obj in selected_shapes():
	split_and_group = apply_action('object-stroke-to-path', obj)
	split = split_and_group[0].ungroup()
	apply_path_operation('difference', split)
```
The extra group doesn't bother me much, but it may be something to look into at some point :)

**Reply by spakin** [reply]
This was a tricky bug to track down, but I believe it's fixed in the latest version on GitHub (0333e3e591ce72a9dfca0b8e2379c4c0b9116a0d).

Note that the order of objects returned by `apply_action` is not guaranteed.  Ideally, your code should search `split_and_group` for an actual group to ungroup.  That said, as a quick hack, the current version of Simple Inkscape Scripting happens to return the group as `split_and_group[-1]`.
